pkg/controller/v1alpha1/llmisvc/workload_single_node.go

```diff
@@ -18,17 +18,176 @@ package llmisvc
+			Selector: &metav1.LabelSelector{
+				MatchLabels: labels,
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: labels,
+				},
+			},
+		},
+	}
+
+	d.Spec.Template.Spec = *llmSvc.Spec.Template.DeepCopy()
+	if hasRoutingSidecar(d.Spec.Template.Spec) {
+		log.FromContext(ctx).Info("Main container has a routing sidecar")
+
+		serviceAccount := r.expectedSingleNodeMainServiceAccount(llmSvc)
+		d.Spec.Template.Spec.ServiceAccountName = serviceAccount.GetName()
+		s := routingSidecar(&d.Spec.Template.Spec)
+		if llmSvc.Spec.Router != nil {
+			s.Env = append(s.Env, corev1.EnvVar{
+				Name:      "INFERENCE_POOL_NAME",
+				Value:     llmSvc.Spec.Router.Scheduler.InferencePoolName(llmSvc),
+				ValueFrom: nil,
+			})
+		}
+	}
+
+	if err := r.attachModelArtifacts(llmSvc, &d.Spec.Template.Spec, storageConfig); err != nil {
+		return nil, fmt.Errorf("failed to attach model artifacts to main deployment: %w", err)
+	}
+
+	log.FromContext(ctx).V(2).Info("Expected main deployment", "deployment", d)
+
+	return d, nil
+}
+
+func (r *LLMISVCReconciler) reconcileSingleNodePrefill(ctx context.Context, llmSvc *v1alpha1.LLMInferenceService) error {
+	prefill := r.expectedPrefillMainDeployment(ctx, llmSvc)
+	if llmSvc.Spec.Prefill == nil {
+		if err := Delete(ctx, r, llmSvc, prefill); err != nil {
+			return fmt.Errorf("failed to delete prefill main deployment: %w", err)
+		}
+		return nil
+	}
+	if err := Reconcile(ctx, r, llmSvc, &appsv1.Deployment{}, prefill, semanticDeploymentIsEqual); err != nil {
+		return fmt.Errorf("failed to reconcile prefill deployment %s/%s: %w", prefill.GetNamespace(), prefill.GetName(), err)
+	}
+	return r.propagateDeploymentStatus(ctx, prefill, llmSvc.MarkPrefillWorkloadReady, llmSvc.MarkPrefillWorkloadNotReady)
+}
+
+func (r *LLMISVCReconciler) expectedPrefillMainDeployment(ctx context.Context, llmSvc *v1alpha1.LLMInferenceService) *appsv1.Deployment {
+	labels := map[string]string{
+		"app.kubernetes.io/component": "llminferenceservice-workload-prefill",
+		"app.kubernetes.io/name":      llmSvc.GetName(),
+		"app.kubernetes.io/part-of":   "llminferenceservice",
+		"kserve.io/component":         "workload",
+		"llm-d.ai/role":               "prefill",
+	}
+
+	d := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      kmeta.ChildName(llmSvc.GetName(), "-kserve-prefill"),
+			Namespace: llmSvc.GetNamespace(),
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(llmSvc, v1alpha1.LLMInferenceServiceGVK),
+			},
+			Labels: labels,
+		},
+	}
+
+	if llmSvc.Spec.Prefill != nil {
+		d.Spec = appsv1.DeploymentSpec{
+			Replicas: llmSvc.Spec.Prefill.Replicas,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: labels,
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: labels,
+				},
+			},
+		}
+
+		if llmSvc.Spec.Prefill.Template != nil {
+			d.Spec.Template.Spec = *llmSvc.Spec.Prefill.Template.DeepCopy()
```

Comment:
Do we expect user to provide the full deployment template? I thought we have the preset configurations

Reply:
in the top-level case `spec.template`, that is the fallback case so even when `template = nil` and no other case is entered in the merge logic

https://github.com/kserve/kserve/blob/64d775bb99a4d7f24837fe01a4c5f0a10853de28/pkg/controller/v1alpha1/llmisvc/config_merge.go#L121-L122